core/src/main/java/com/linecorp/armeria/client/HttpResponseDecoder.java

```diff
@@ -127,16 +134,24 @@ final void failUnfinishedResponses(Throwable cause) {
         }
     }
 
+    @Nullable
+    abstract KeepAliveHandler keepAliveHandler();
+
     final void disconnectWhenFinished() {
         disconnectWhenFinished = true;
     }
 
     final boolean needsToDisconnectNow() {
-        return disconnectWhenFinished && !hasUnfinishedResponses();
+        return needsToDisconnectWhenFinished() && !hasUnfinishedResponses();
     }
 
     final boolean needsToDisconnectWhenFinished() {
-        return disconnectWhenFinished;
+        if (disconnectWhenFinished) {
```

Comment:
Would it be worth to try to make `KeepAliveHandler` not nullable and handle all keep-aliving? 🤔

Reply:
I also think about that. KeepAliveHandler works in most cases unless users disable both an idle timeout and a request or response timeout. That would be a rare case.